### PR TITLE
Support Gateway plugin LLM provider

### PR DESCRIPTION
**Pull request by gabrielfu**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/12611?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12611/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12611
```

</p>
</details>

### Related Issues/PRs

replaces #12116

### What changes are proposed in this pull request?

Users can create an installable plugin package that integrates any non-native LLM provider to MLflow Gateway (and Deployments Servers).

They need to implements the provider and config classes, and declare the provider name in the `project.entry-points."mlflow.gateway.providers"` Python entry point. Then, they can specify the provider in the config.yaml file.

See docs `llms/deployments/index.html#plugin-llm-provider-experimental` and `examples/deployments/deployments_servers/plugin` for more information.

per https://github.com/mlflow/mlflow/pull/12573#issuecomment-2211093040, we will continue using the base classes from Gateway module

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [X] Manual tests

See examples.

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [X] Yes. I've updated:
  - [X] Examples
  - [ ] API references
  - [X] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Support plugin LLM provider in MLflow Deployments Server

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by harupy** [comment]
> Also, we might need to declare @developer_stable to those base classes?

+1

File: examples/deployments/deployments_server/plugin/foo-provider/foo/providers.py
Comment (harupy) [inline examples/deployments/deployments_server/plugin/foo-provider/foo/providers.py:1]: rename foo -> my_llm?

File: mlflow/gateway/providers/base.py [hunk 1]
```diff
@@ -14,6 +15,7 @@ class BaseProvider(ABC):
 
     NAME: str = ""
     SUPPORTED_ROUTE_TYPES: Tuple[str, ...]
+    CONFIG_TYPE: Type[ConfigModel] = None
```
Comment (harupy) [inline mlflow/gateway/providers/base.py:18]:
```suggestion
    CONFIG_TYPE: Type[ConfigModel]
```

Do we need `= None`?

File: docs/source/llms/deployments/index.rst [hunk 2]
```diff
@@ -1182,6 +1183,139 @@ Here are some examples for how you might use curl to interact with the MLflow De
...
+and a config class that inherits from ``mlflow.gateway.base_models.ConfigModel``.
+
+.. code-block:: python
+    :caption: Example
+
+    from typing import AsyncIterable
+
+    from pydantic import validator
+    from mlflow.deployments.server.base_models import ConfigModel
+    from mlflow.deployments.server.config import RouteConfig, _resolve_api_key_from_input
```
Comment (harupy) [inline docs/source/llms/deployments/index.rst:1208]: Is it possible to avoid using `_resolve_api_key_from_input` (private function)?

**Review by harupy — APPROVED** [review]
Looks good to me! @BenWilson2 WDYT?

File: docs/source/llms/deployments/index.rst [hunk 3]
```diff
@@ -1182,6 +1183,139 @@ Here are some examples for how you might use curl to interact with the MLflow De
...
+
+
+    class MyLLMConfig(ConfigModel):
+        # This model defines the configuration for the provider such as API keys
+        my_llm_api_key: str
+
+        @validator("my_llm_api_key", pre=True)
+        def validate_my_llm_api_key(cls, value):
+            # This resolves the API key from an environment variable
+            return _resolve_api_key_from_input(value)
```
Comment (harupy) [inline docs/source/llms/deployments/index.rst:1220]:
```suggestion
            return os.environ[value]
```

File: mlflow/gateway/providers/base.py (hunk 1 above)
Comment (gabrielfu) [reply mlflow/gateway/providers/base.py]: removed it

File: docs/source/llms/deployments/index.rst [hunk 4]
```diff
@@ -1182,6 +1183,139 @@ Here are some examples for how you might use curl to interact with the MLflow De
...
+
+MLflow already provides a number of providers by default. Your plugin name cannot be the same as any one
+of them. See :ref:`deployments_configuration_details` for a complete list of default providers.
+
+Finally, you need to install the plugin package in the same environment as the MLflow Deployments Server.
+
+.. important::
+
+    Only install plugin packages you trust. Starting a server with plugin provider will
+    execute arbitrary codes in the plugin package.
```
Comment (BenWilson2) [inline docs/source/llms/deployments/index.rst:1297]:
```warning
    Only install plugin packages from sources that you trust. Starting a server with a plugin provider will
    execute any arbitrary code that is defined within the plugin package.
```

File: mlflow/gateway/provider_registry.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
+from typing import Dict, Type, Union
+
+from mlflow import MlflowException
+from mlflow.gateway.config import Provider
+from mlflow.gateway.providers import BaseProvider
+from mlflow.utils.os import get_entry_points
+
+
+class ProviderRegistry:
```
Comment (BenWilson2) [inline mlflow/gateway/provider_registry.py:9]: Love this

**Review by BenWilson2 — APPROVED** [review]
Tiny nit on the warning statement wording. Otherwise, this looks great! 
I think this is a very reasonable approach to extend plugin support for users and isn't too overly complex. 
We should definitely mark the ABC as dev stable as a reminder for us to not modify the structure in such a way that would break this plugin functionality.

**Pull request merged**